node-repository/src/main/java/com/yahoo/vespa/hosted/provision/NodeRepository.java

```diff
@@ -921,13 +924,58 @@ public Mutex lock(ApplicationId application, Duration timeout) {
     /** Create a lock which provides exclusive rights to modifying unallocated nodes */
     public Mutex lockUnallocated() { return db.lockInactive(); }
 
-    /** Acquires the appropriate lock for this node */
-    public Mutex lock(Node node) {
+    /** Returns the unallocated/application lock, and the node acquired under that lock. */
+    public Optional<NodeMutex> lockAndGet(Node node) {
+        Node staleNode = node;
+
+        for (int i = 0; i < 4; ++i) {
+            Mutex lockToClose = lock(staleNode);
+            try {
+                Optional<Node> freshNode = getNode(staleNode.hostname(), staleNode.state());
+                if (freshNode.isEmpty()) {
+                    freshNode = getNode(staleNode.hostname());
+                    if (freshNode.isEmpty()) {
+                        return Optional.empty();
+                    }
+                }
+
+                if (Objects.equals(freshNode.get().allocation().map(Allocation::owner),
+                                   staleNode.allocation().map(Allocation::owner))) {
+                    NodeMutex nodeMutex = new NodeMutex(freshNode.get(), lockToClose);
+                    lockToClose = null;
+                    return Optional.of(nodeMutex);
+                }
+
+                // The wrong lock was held when the fresh node was fetched, so try again
+                staleNode = freshNode.get();
+            } finally {
+                if (lockToClose != null) lockToClose.close();
+            }
+        }
+
+        throw new IllegalStateException("Giving up trying to fetch an up to date node under lock: " + node.hostname());
+    }
+
+    /** Returns the unallocated/application lock, and the node acquired under that lock. */
+    public Optional<NodeMutex> lockAndGet(String hostname) {
+        return getNode(hostname).flatMap(this::lockAndGet);
+    }
+
+    /** Returns the unallocated/application lock, and the node acquired under that lock. */
+    public NodeMutex lockAndGetRequired(Node node) {
```

Comment:
Consider naming this `lockAndRequire`.

Reply:
Hm, I like the information `Get` adds, and that these related methods all contains the same `lockAndGet` prefix.